js/pattern.js

```diff
@@ -544,15 +544,19 @@ function initPattern(globals){
 
     function processFold(fold, returnCreaseParams){
 
-        rawFold = JSON.parse(JSON.stringify(fold));//save pre-triangulated for for save later
-        //make 3d
-        for (var i=0;i<rawFold.vertices_coords.length;i++){
-            var vertex = rawFold.vertices_coords[i];
-            if (vertex.length === 2) {//make vertices_coords 3d
-                rawFold.vertices_coords[i] = [vertex[0], 0, vertex[1]];
-            }
+        //add missing coordinates to make 3d, and flip y and z for internal use
+        //FOLD spec says that, beyond two dimensions,
+        //"all unspecified coordinates are implicitly zero"
+        var is2d = true;
+        for (var i=0;i<fold.vertices_coords.length;i++){
+            var vertex = fold.vertices_coords[i];
+            if (vertex[2]) is2d = false;
+            fold.vertices_coords[i] = [vertex[0], vertex[2] || 0, vertex[1]];
```

Comment:
@edemaine 

This line is problematic. See the following Tetrahedron FOLD model:

[tertahedron.fold.zip](https://github.com/amandaghassaei/OrigamiSimulator/files/5611471/tertahedron.fold.zip)

When you open it inf FOLD viewer you can see the faces are oriented correctly (cyan). Also when you load it in OrigamiSimulator:master - the pink faces are outwards.

However, when you open it on the current branch, the "pink" faces are facing inward. This does cause troubles for models like the Dodecahedron as well, but easier to verify with the Tetrahedron. I verified hat removing these lines and returning to:

```
for (var i=0;i<rawFold.vertices_coords.length;i++){
       var vertex = rawFold.vertices_coords[i];
        if (vertex.length === 2) {//make vertices_coords 3d
          is2d = true;
          rawFold.vertices_coords[i] = [vertex[0], 0, vertex[1]];
        }
      }
```

Fixes the problem. Do we really want to always flip the FOLD y and z? it will work as long as FOLD are exported from OrigamiSimulator, but will not be compatible with other apps (e.g. edge unfolder) that export FOLDs and assume something about the dimensions (e.g. closed models with faces in/out)

Reply:
I've always found it disconcerting (and unjustifiable) that we swap y and z in 2D files but not 3D files.  I believe swapping y and z is just for the purposes of rendering, but I guess it affects triangulation somehow too, though I don't see why it would.  Thanks for finding a case where it's a problem -- I'll have to take a look at why.

Reply:
OK, sleeping on this, I see why: swapping y/z flips the coordinate system between left- and right-handed, effectively reversing the notions of cw/ccw. This isn't an issue in 2D, because there it would be the same as mapping (x, y, z) &rarr; (x, -z, y) (because z=0) which doesn't flip handedness, but it's a problem in 3D.

I'm not a fan of this 2D to 3D mapping, but it's historical, and fixing this really isn't an issue for this PR. So I've reverted that change, but kept the safer check for `is2d` (*all* vertices need to be 2D, not just one).

So, please test again!